### PR TITLE
Enable server side diff for comparing resources

### LINKED ISSUE #200: Enable serverSideDiff option when comparing diff
# Summary

As we are using server side apply by default, we should enable serverSideDiff [option](https://github.com/numaproj-labs/numaplane/blob/main/internal/sync/diff.go#L205) as well when comparing diff.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by chandankumar4**
Fixes #200 

### Modifications
- Enable serverSideDiff Option when comparing diff as we are using serverSideApply.

### Verification

Tested in local k8s cluster, these are dry-run logs
```bash
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.548199+05:30","msg":"Applying resource ConfigMap/numaflow-controller-config in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.555639+05:30","msg":"Applying resource Deployment/numaflow-controller in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.5637+05:30","msg":"Applying resource InterStepBufferService/default in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.570174+05:30","msg":"Applying resource Role/numaflow-role in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.575554+05:30","msg":"Applying resource RoleBinding/numaflow-role-binding in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.579973+05:30","msg":"Applying resource Pipeline/simple-pipeline in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","logger":"numaplane","caller":"kube/resource_ops.go:251","ts":"2024-04-17T14:51:42.586198+05:30","msg":"Applying resource ServiceAccount/numaflow-sa in cluster: https://127.0.0.1:49383, namespace: numaflow-pipeline"}
{"level":"info","worker":15,"gitSyncKey":"numaplane-system/gitsync-example","gitsync":"numaplane-system/gitsync-example","logger":"numaplane.synchronizer","caller":"sync/sync_context.go:387","ts":"2024-04-17T14:51:42.590869+05:30","msg":"Syncing"}
{"level":"info","worker":15,"gitSyncKey":"numaplane-system/gitsync-example","gitsync":"numaplane-system/gitsync-example","logger":"numaplane.synchronizer","caller":"sync/sync_context.go:409","ts":"2024-04-17T14:51:42.597809+05:30","msg":"Tasks (dry-run)"}
{"level":"info","worker":15,"gitSyncKey":"numaplane-system/gitsync-example","gitsync":"numaplane-system/gitsync-example","logger":"numaplane.synchronizer","caller":"sync/sync_context.go:921","ts":"2024-04-17T14:51:42.597863+05:30","msg":"Updating operation state. phase:  -> Succeeded, message: '' -> 'successfully synced (no more tasks)'"}
{"level":"info","worker":15,"gitSyncKey":"numaplane-system/gitsync-example","logger":"numaplane.synchronizer","caller":"sync/syncer.go:245","ts":"2024-04-17T14:51:42.598378+05:30","msg":"GitSync object is successfully synced."}

```

### CODE REVIEW

File: internal/sync/syncer.go [hunk 1]
```diff
@@ -343,8 +344,18 @@ func (s *Syncer) compareState(gitSync *v1alpha1.GitSync, targetObjs []*unstructu
...
+		log.Errorf("CompareAppState error getting resource operations: %s", err)
+	}
+	defer cleanup()
+
 	diffOpts := []diff.Option{
 		diff.WithLogr(*logger.New().WithValues("gitsync", fmt.Sprintf("%s/%s", gitSync.Namespace, gitSync.Name)).LogrLogger),
+		diff.WithServerSideDiff(true),
+		diff.WithServerSideDryRunner(diff.NewK8sServerSideDryRunner(resourceOps)),
+		diff.WithManager(common.SSAManager),
+		diff.WithGVKParser(clusterCache.GetGVKParser()),
```
Comment (xdevxy) [inline internal/sync/syncer.go:358]: Why do we need this?

File: internal/sync/syncer.go [hunk 2]
```diff
@@ -355,13 +366,31 @@ func (s *Syncer) compareState(gitSync *v1alpha1.GitSync, targetObjs []*unstructu
...
+}
+
+func applyAnnotationAndNamespace(manifests []*unstructured.Unstructured, gitSyncName, destinationNamespace string) ([]*unstructured.Unstructured, error) {
 	uns := make([]*unstructured.Unstructured, 0)
 	for _, m := range manifests {
 		err := kubernetes.SetGitSyncInstanceAnnotation(m, common.AnnotationKeyGitSyncInstance, gitSyncName)
 		if err != nil {
 			return nil, err
 		}
+		m.SetNamespace(destinationNamespace)
```
Comment (xdevxy) [inline internal/sync/syncer.go:393]: Is this just for extra safety? Or we have a bug without it? Shouldn't when we set the namespace in syncing taking care of it?

Comment (chandankumar4) [reply internal/sync/syncer.go:393]: After enabling server side diff, without knowing the namespace it was not able to find the resource in cluster for diff. that's why we need to set it upfront

File: internal/sync/syncer.go (hunk 1 above)
Comment (chandankumar4) [reply internal/sync/syncer.go:358]: without `WithGVKParser` serverSide diff was not working

**Review by xdevxy — APPROVED** [review]

**Pull request merged**